tasks/us/hierarchy.py

```diff
@@ -0,0 +1,332 @@
+                        'REFERENCES {info_table} (geoid, level) '.format(
+            rel_table=self.input()['rel'].qualified_tablename,
+            info_table=self.input()['info'].qualified_tablename))
+        session.commit()
+
+    def complete(self):
+        session = current_session()
+        sql = "SELECT 1 FROM information_schema.constraint_column_usage " \
+              "WHERE table_schema = '{schema}' " \
+              "  AND table_name = '{table}' " \
+              "  AND constraint_name = '{constraint}'"
+        table = self.input()['info']
+        check = sql.format(schema=table.schema,
+                           table=table.tablename,
+                           constraint='ushierarchy_fk_parent')
+        result = session.execute(check).fetchall()
+        return len(result) > 0
+
+
+class _YearLevelsTask:
+    year = IntParameter()
+    levels = ListParameter(significant=False)
+
+
+class USHierarchyInfoUnion(TempTableTask, _YearLevelsTask):
+
+    def requires(self):
+        return [USLevelInfo(year=self.year, geography=level)
+                for level in self.levels]
+
+    def run(self):
+        session = current_session()
+        tablename = self.output().qualified_tablename
+        session.execute(_union(self.input(), tablename))
+        alter_sql = 'ALTER TABLE {tablename} ADD PRIMARY KEY (geoid, level)'
+        session.execute(alter_sql.format(tablename=tablename))
+        session.commit()
+
+
+class USHierarchyChildParentsUnion(TempTableTask, _YearLevelsTask):
+
+    def requires(self):
+        child_parents = self._child_parents()
+        return [USLevelInclusionHierarchy(year=self.year,
+                                          current_geography=child_parent[0],
+                                          parent_geography=child_parent[1])
+                for child_parent in child_parents]
+
+    def _child_parents(self):
+        child_parents = []
+        previous = None
+        for idx, level in enumerate(self.levels):
+            if previous:
+                child_parents.append([level, previous])
+            previous = level
+        return child_parents
+
+    def run(self):
+        session = current_session()
+        tablename = self.output().qualified_tablename
+        session.execute(_union(self.input(), tablename))
+        alter_sql = 'ALTER TABLE {tablename} ADD PRIMARY KEY ' \
+                    '(child_id, child_level, parent_id, parent_level)'
+        session.execute(alter_sql.format(tablename=tablename))
+        session.commit()
+
+
+GEOGRAPHIES_ABBREVIATIONS = {
+    'school_district_secondary': 'sds',
+    'school_district_unified': 'sdu',
+    'school_district_elementary': 'sde'
+}
+
+
+def _abbreviation(geography):
+    return GEOGRAPHIES_ABBREVIATIONS.get(geography, geography)
+
+
+def abbr_tablename(target, geographies, year):
+    abbrs = [_abbreviation(geography) for geography in geographies]
+
+    if [x for x in zip(abbrs, geographies) if x[0] != x[1]]:
```

Comment:
This is hard to read (abbreviations + zip). It made me do a double take. 

Is this `if` even needed? This is to avoid the case where there are no abbreviations. Is there a problem if you replace the name even when no geography names are abbreviated?

Reply:
I just wanted to avoid renaming tables if it's not needed. I don't have strong opinions on this.